servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/DefaultNettyConnection.java

```diff
@@ -268,6 +275,49 @@ protected void handleSubscribe(
         };
     }
 
+    private Publisher<Read> registerReadObserver(final Publisher<Read> readPublisher) {
+        return defer(() -> {
+            final DataObserver dataObserver = this.dataObserver;
+            final ReadObserver observer;
+            if (dataObserver == null ||
+                    (observer = safeReport(dataObserver::onNewRead, dataObserver, "new read")) == null) {
+                return readPublisher.subscribeShareContext();
```

Comment:
Prefer operators over new source creation and context propagation handling complexity.

```java
        return readPublisher.liftSync(target -> {
            final DataObserver dataObserver = this.dataObserver;
            final ReadObserver observer;
            if (dataObserver == null ||
                    (observer = safeReport(dataObserver::onNewRead, dataObserver, "new read")) == null) {
                return target;
            }
            return new PublisherSource.Subscriber<Read>() {
                @Override
                public void onSubscribe(final Subscription subscription) {
                    target.onSubscribe(new Subscription() {
                        @Override
                        public void request(final long n) {
                            safeReport(() -> observer.requestedToRead(n), observer, "requested to read");
                            subscription.request(n);
                        }

                        @Override
                        public void cancel() {
                            safeReport(observer::readCancelled, observer, "read cancelled");
                            subscription.cancel();
                        }
                    });
                }

                @Override
                public void onNext(@Nullable final Read read) {
                    safeReport(observer::itemRead, observer, "item read");
                    target.onNext(read);
                }

                @Override
                public void onError(final Throwable t) {
                    safeReport(() -> observer.readFailed(t), observer, "read failed", t);
                    target.onError(t);
                }

                @Override
                public void onComplete() {
                    safeReport(observer::readComplete, observer, "read complete");
                    target.onComplete();
                }
            };
        });

```